ext/curl/interface.c

```diff
@@ -2095,13 +2114,249 @@ void _php_setup_easy_copy_handlers(php_curl *ch, php_curl *source)
 	(*source->clone)++;
 }
 
+#if LIBCURL_VERSION_NUM >= 0x073800 /* 7.56.0 */
+static size_t read_cb(char *buffer, size_t size, size_t nitems, void *arg) /* {{{ */
+{
+	struct mime_data_cb_arg *cb_arg = (struct mime_data_cb_arg *) arg;
+	ssize_t numread;
+
+	ZEND_ASSERT(!cb_arg->ch);
+
+	if (cb_arg->stream == NULL) {
+		if (!(cb_arg->stream = php_stream_open_wrapper(ZSTR_VAL(cb_arg->filename), "rb", IGNORE_PATH, NULL))) {
+			return CURL_READFUNC_ABORT;
+		}
+	}
+	numread = php_stream_read(cb_arg->stream, buffer, nitems * size);
+	if (numread < 0) {
+		php_stream_close(cb_arg->stream);
+		cb_arg->stream = NULL;
+	}
+	return (numread >= 0) ? numread : CURL_READFUNC_ABORT;
+}
+/* }}} */
+
+static int seek_cb(void *arg, curl_off_t offset, int origin) /* {{{ */
+{
+	struct mime_data_cb_arg *cb_arg = (struct mime_data_cb_arg *) arg;
+	int res;
+
+	ZEND_ASSERT(!cb_arg->ch);
+
+	if (cb_arg->stream == NULL) {
+		if (!(cb_arg->stream = php_stream_open_wrapper(ZSTR_VAL(cb_arg->filename), "rb", IGNORE_PATH, NULL))) {
+			return CURL_SEEKFUNC_CANTSEEK;
+		}
+	}
+	res = php_stream_seek(cb_arg->stream, offset, origin);
+	return res == SUCCESS ? CURL_SEEKFUNC_OK : CURL_SEEKFUNC_CANTSEEK;
+}
+/* }}} */
+
+static void free_cb(void *arg) /* {{{ */
+{
+	struct mime_data_cb_arg *cb_arg = (struct mime_data_cb_arg *) arg;
+
+	ZEND_ASSERT(!cb_arg->ch);
+
+	if (cb_arg->stream != NULL) {
+		php_stream_close(cb_arg->stream);
+		cb_arg->stream = NULL;
+	}
+}
+/* }}} */
+
+static zval *get_postfields_of_handle(php_curl *ch) /* {{{ */
+{
+	struct mime_data_cb_arg *cb_arg, **cb_arg_p;
+
+	cb_arg_p = zend_llist_get_last(&ch->to_free->stream);
+	while (cb_arg_p) {
+		cb_arg = *cb_arg_p;
+		if (cb_arg->ch == ch) {
+			return &cb_arg->postfields;
+		}
+		cb_arg_p = zend_llist_get_prev(&ch->to_free->stream);
+	}
+	return NULL;
+}
+/* }}} */
+#endif
+
+static int build_mime_structure_from_hash(php_curl *ch, zval *zpostfields) /* }}} */
```

Comment:
Please precommit function extraction.

Reply:
Done with a6d385a8e8d6955214640a954577dd9e872dfb69. I'll rebase this PR ASAP.

Reply:
Squashed and rebased.